specs/core/1_shard-data-chains.md

```diff
@@ -95,517 +91,307 @@ We define the following Python custom types for type hinting and readability:
 
 | Name | Value |
 | - | - |
-| `BASEFEE_ADJUSTMENT_FACTOR` | `2**3` (= 8) |
-| `REWARD_COEFFICIENT_BASE` | `2**20` ( = 1,048,576) |
+| `BLOCK_BODY_PRICE_QUOTIENT` | `2**3` (= 8) |
 
 ### Signature domain types
 
-The following types are defined, mapping into `DomainType` (little endian):
-
 | Name | Value |
 | - | - |
 | `DOMAIN_SHARD_PROPOSER` | `128` |
 | `DOMAIN_SHARD_ATTESTER` | `129` |
 
-### TODO PLACEHOLDER
-
-| Name | Value |
-| - | - |
-| `PLACEHOLDER` | `2**3` |
-
-## Data structures
-
-_Note: the shard block header structure is carefully designed so that all of the values have the same depth in a hash tree implementation, so `hash_tree_root(SSZ_partial(x)) == hash_tree_root(x)` (using the "left-to-right leaves" scheme [here](https://github.com/ethereum/eth2.0-specs/issues/1303)), which allows shard block headers to look like an SSZ object when in the crosslink structure. This is done by balancing it so that 7 or 8 items are on the left side (the "core") and two 96-byte (ie. 3*2 = 6 chunk) items are on the right side. Change with care._
-
-### `ShardBlockHeader`
-
-```python
-class ShardBlockHeader(Container):
-    core: ShardBlockCore
-    signatures: ShardBlockSignatures
-```
+## Containers
 
 ### `ShardBlock`
 
 ```python
-class ShardBlock(Container):
-    core: ExtendedShardBlockCore
-    signatures: ShardBlockSignatures
-```
-
-### `ShardBlockSignatures`
-
-```python
-class ShardBlockSignatures(Container):
-    attestation_signature: BLSSignature
-    proposer_signature: BLSSignature
-```
-
-### `ShardBlockCore`
-
-```python
-class ShardBlockCore(Container):
+class ShardBlock(FlatContainer):
```

Comment:
This is actually a (big) problem with this approach. With a flat container, you no longer have this nice property that an object with an entire subtree substituted with its root necessarily has the same hash tree root as the original object, so you can't have shard blocks and shard block headers in the same way. Not sure how to get around this within the FlatContainer model.

This is arguably one reason why not doing flat containers and instead taking Danny's suggestion and having "embedded vectors (and containers)" would be better. Basically, the idea would be that if you have a container as follows:

```python
class Foo(Container):
    bar: uint64
    baz: Embedded[Vector[uint64, 4]]
    bav: Bytes32
```

It would be treated identically to:

```python
class Foo(Container):
    bar: uint64
    baz_0: uint64
    baz_1: uint64
    baz_2: uint64
    baz_3: uint64
    bav: Bytes32
```

This way you avoid introducing a separate type of container and instead have flattening as a feature that you can use or not use for specific elements at will.

Reply:
Was just about to question if we indeed get this property anymore. The embedded looks like a reasonable approach

Reply:
So are we going to replace FlatContainer with Embedded here? I guess the `attestations` and `signature` would need to be part of a 2-item Embedded container where the items are themselves Embedded to get flatness while gaining the ability to use the standard signature root method?